File: dataset/output-bk/GEMINI_gemini-1.5-pro-001/Task181_GEMINI_gemini-1.5-pro-001.java
import java.util.concurrent.locks.ReentrantLock;

public class BankAccount {
    private double balance;
    private final ReentrantLock lock = new ReentrantLock();
    private boolean isOpen;

    public BankAccount(double initialBalance) {
        this.balance = initialBalance;
        this.isOpen = true;
    }

    public void open() {
        if (isOpen) {
            throw new IllegalStateException("Account already open");
        }
        isOpen = true;
    }

    public void close() {
        if (!isOpen) {
            throw new IllegalStateException("Account already closed");
        }
        isOpen = false;
    }

    public void deposit(double amount) {
        if (!isOpen) {
            throw new IllegalStateException("Account closed");
        }
        if (amount <= 0) {
            throw new IllegalArgumentException("Deposit amount must be positive");
        }
        lock.lock();
        try {
            balance += amount;
        } finally {
            lock.unlock();
        }
    }

    public void withdraw(double amount) {
        if (!isOpen) {
            throw new IllegalStateException("Account closed");
        }
        if (amount <= 0) {
            throw new IllegalArgumentException("Withdrawal amount must be positive");
        }
        lock.lock();
        try {
            if (balance < amount) {
                throw new IllegalArgumentException("Insufficient funds");
            }
            balance -= amount;
        } finally {
            lock.unlock();
        }
    }
}

File: dataset/output/GEMINI_gemini-1.5-pro-001/Task181_GEMINI_gemini-1.5-pro-001.cpp
#include <iostream>
#include <mutex>

class BankAccount {
private:
    double balance;
    std::mutex mutex;
    bool isOpen;

public:
    BankAccount(double initialBalance = 0) : balance(initialBalance), isOpen(true) {}

    void open() {
        std::lock_guard<std::mutex> lock(mutex);
        if (!isOpen) {
            isOpen = true;
        } else {
            throw std::runtime_error("Account is already open.");
        }
    }

    void close() {
        std::lock_guard<std::mutex> lock(mutex);
        if (isOpen) {
            isOpen = false;
        } else {
            throw std::runtime_error("Account is already closed.");
        }
    }

    void deposit(double amount) {
        std::lock_guard<std::mutex> lock(mutex);
        if (!isOpen) {
            throw std::runtime_error("Cannot deposit to a closed account.");
        }
        if (amount <= 0) {
            throw std::invalid_argument("Deposit amount must be positive.");
        }
        balance += amount;
    }

    void withdraw(double amount) {
        std::lock_guard<std::mutex> lock(mutex);
        if (!isOpen) {
            throw std::runtime_error("Cannot withdraw from a closed account.");
        }
        if (amount <= 0) {
            throw std::invalid_argument("Withdrawal amount must be positive.");
        }
        if (balance < amount) {
            throw std::runtime_error("Insufficient funds.");
        }
        balance -= amount;
    }

    double getBalance() {
        std::lock_guard<std::mutex> lock(mutex);
        if (!isOpen) {
            throw std::runtime_error("Cannot get balance of a closed account.");
        }
        return balance;
    }
};